src/companion.rs

```diff
@@ -1,195 +1,233 @@
 use regex::Regex;
 use snafu::ResultExt;
-use tokio::process::Command;
+use std::path::Path;
 
-use crate::{error::*, github_bot::GithubBot, Result};
+use crate::{cmd::*, error::*, github_bot::GithubBot, Result};
 
 pub async fn companion_update(
 	github_bot: &GithubBot,
-	base_owner: &str,
-	base_repo: &str,
-	head_owner: &str,
-	head_repo: &str,
-	branch: &str,
-) -> Result<Option<String>> {
-	let res = companion_update_inner(
-		github_bot, base_owner, base_repo, head_owner, head_repo, branch,
+	owner: &str,
+	owner_repo: &str,
+	contributor: &str,
+	contributor_repo: &str,
+	contributor_branch: &str,
+) -> Result<String> {
+	let token = github_bot.client.auth_key().await?;
+	let secrets_to_hide = [token.as_str()];
+	let secrets_to_hide = Some(&secrets_to_hide[..]);
+
+	let owner_repository_domain =
+		format!("github.com/{}/{}.git", owner, owner_repo);
+	let owner_remote_address = format!(
+		"https://x-access-token:{}@{}",
+		token, owner_repository_domain
+	);
+	let repo_dir = format!("./{}", owner_repo);
+
+	if Path::new(&repo_dir).exists() {
+		log::info!("{} is already cloned; skipping", owner_repository_domain);
+	} else {
+		run_cmd_in_cwd(
+			"git",
+			&["clone", "-v", &owner_remote_address],
+			CommandMessage::Configured(CommandMessageConfiguration {
+				secrets_to_hide,
+				are_errors_silenced: false,
+			}),
+		)
+		.await?;
+	}
+
+	let contributor_repository_domain =
+		format!("github.com/{}/{}.git", contributor, contributor_repo);
+	let contributor_remote_branch =
+		format!("{}/{}", contributor, contributor_branch);
+	let contributor_remote_address = format!(
+		"https://x-access-token:{}@{}",
+		token, contributor_repository_domain
+	);
+
+	// The contributor's remote entry might exist from a previous run (not expected for a fresh
+	// clone). If so, delete it so that it can be recreated.
+	if run_cmd(
+		"git",
+		&["remote", "get-url", contributor],
+		&repo_dir,
+		CommandMessage::Configured(CommandMessageConfiguration {
+			secrets_to_hide,
+			are_errors_silenced: true,
+		}),
+	)
+	.await
+	.is_ok()
+	{
+		run_cmd(
+			"git",
+			&["remote", "remove", contributor],
+			&repo_dir,
+			CommandMessage::Configured(CommandMessageConfiguration {
+				secrets_to_hide,
+				are_errors_silenced: false,
+			}),
+		)
+		.await?;
+	}
+	run_cmd(
+		"git",
+		&["remote", "add", contributor, &contributor_remote_address],
```

Comment:
This could be any random user/org and repository, not necessarily inside `paritytech`, right? In fact, I can't think of a reason for someone to clone `paritytech/polkadot` to `paritytech/foobar` and then submit PRs from that to `paritytech/polkadot`.

I vaguely remember that this whole bot-driven companion workflow cannot be used for PRs from forks, because those won't be run in CI or something like that. Is that still the case? Or did I remember that wrong?

Looking at the code that is deleted in this PR, it seems to have created a remote called `temp` which also points to the "owner" repo (so e.g. `paritytech/polkadot`?). I don't understand why it was done that way. Whatever the reason, the new code creates a remote pointing at a 3rd party repo (so e.g. `bob/polkadot`), named after the user or organization where a PR originates. And further down (line 201) new commits are pushed to that origin. In what scenarios would that succeed or fail? The bot usually doesn't have push permissions to `bob/polkadot`, right?

Reply:
> This could be any random user/org and repository, not necessarily inside paritytech, right?

That's right. There's nothing blocking it from being started for any organization, although the expect scenario is a Substrate PR trying a companion update on Polkadot; in fact, it's hardcoded to only work for Substrate.

https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/webhook.rs#L1205

This code assumes that the bot will be installed in the companion PR's repository (`polkadot` in this case) since, as you noted, it needs push permissions to push commits to the companion.

---

> I vaguely remember that this whole bot-driven companion workflow cannot be used for PRs from forks, because those won't be run in CI or something like that. Is that still the case? Or did I remember that wrong?

I would expect the merge and push to work granted the fork's PR has [enabled edits from maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) since the bot should have the relevant permissions for being classified as a "maintainer".

For CI [the bot simply checks that everything is green](https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/webhook.rs#L252) without regards from where those statuses came from, therefore I would expect it to work fine even if the branch does not belong to a `paritytech` repository as CI also runs on PRs from forks.

To answer the question more directly: this scenario is not taken into account currently, nor tested, so there's no certain response.

---

> Looking at the code that is deleted in this PR, it seems to have created a remote called temp which also points to the "owner"

`temp` was what's now called the contributor's MR. In the case of e.g. https://github.com/paritytech/polkadot/pull/2610 it's `paritytech/polkadot:gav-abstract-payout-curve` (you can hover over the name to see it).

![image](https://user-images.githubusercontent.com/77391175/111668731-cf9ee880-87f4-11eb-932f-6aa4f1c44d74.png)

Failure or success is related to the bot having push permissions **to the PR**, not necessarily the contributor's repository. In that case the branch was created directly in `paritytech/polkadot`, where the bot is assumed to be installed, so it'll always succeed in that regard. I expect it to also work for forks' PRs granted they have the enabled "edits from maintainers" as I mentioned above, although that's really not taken into account currently.

Effectively the refactored code should achieve the same as the deleted code; most noticeably the naming has been made more meaningful (from `temp` to `contributor`) and made easier to follow.

> I don't understand why it was done that way

The point of this flow is twofold

- Create a master merge commit
- Update `Cargo.lock` to the latest SHA on Github after the companion on Substrate is merged

The bot has to commit those changes on the PR's author behalf. `temp` was pointing at the PR author's (here called the "contributor") repository and the new code should still be doing the same; it *just happens* that the "owner" and "contributor" are both `paritytech` in the case of e.g. https://github.com/paritytech/polkadot/pull/2610 but there's nothing in the code stopping it from running it on forks.